chia/wallet/puzzles/curry.clib

```diff
@@ -0,0 +1,95 @@
+  ; client code should call the "build table" code once, then pass it around to anyone that wants to
+  ; call `curry` or `curry2`. This is pretty intrusive, so for now we'll just use the existing
+  ; global constant infrastructure, and include it as a fixed table so the tree of four values will
+  ; appear in all code that includes this file, and it will compress better in generators.
+
+  (defun-inline sha256_one _noargs (f (f constant-tree)))
+  (defun-inline sha256_one_one _noargs (r (f constant-tree)))
+  (defun-inline two_sha256_one_a_kw _noargs (f (r constant-tree)))
+  (defun-inline two_sha256_one_c_kw _noargs (r (r constant-tree)))
+
+  ;; this returns the sha256 tree hash of expression F = `((q . a1) a2)`
+  (defun hash-expression-F (a1 a2)
+         (sha256 TWO (sha256 TWO (sha256_one_one) a1)
+                 (sha256 TWO a2 (sha256_one)))
+  )
+
+  ;; Given the tree hash `environment-hash` of an environment tree E
+  ;; and the tree hash `parameter-hash` of a constant parameter P
+  ;; return the tree hash of the tree corresponding to
+  ;; `(c (q . P) E)`
+  ;; This is the new environment tree with the addition parameter P curried in.
+  ;;
+  ;; Note that `(c (q . P) E)` = `(c . ((q . P) . (E . 0)))`
+
+  (defun-inline update-hash-for-parameter-hash (parameter-hash environment-hash)
+     (sha256 (two_sha256_one_c_kw) (hash-expression-F parameter-hash environment-hash))
+  )
+
+  ;; Given the tree hash `environment-hash` of an environment tree E
+  ;; and the tree hash `mod-hash` of a mod M
+  ;; return the tree hash of the tree corresponding to
+  ;; `(a (q . M) E)`
+  ;; This is the hash of a new function that adopts the new environment E.
+  ;; This is used to build of the tree hash of a curried function.
+  ;;
+  ;; Note that `(a (q . M) E)` = `(a . ((q . M)  . (E . 0)))`
+
+  (defun-inline tree-hash-of-apply (mod-hash environment-hash)
+     (sha256 (two_sha256_one_a_kw) (hash-expression-F mod-hash environment-hash))
+  )
+
+  ;; This function recursively calls `update-hash-for-parameter-hash`
+
+  (defun calculate-hash-of-curried-parameters (curry-parameter-hashes)
+     (if curry-parameter-hashes
+      (update-hash-for-parameter-hash (f curry-parameter-hashes) (calculate-hash-of-curried-parameters (r curry-parameter-hashes)))
+      (sha256_one_one)
+     )
+  )
+
+  ;; mod-hash:
+  ;;   the hash of a puzzle function, ie. a `mod`
+  ;;
+  ;; curry-parameter-hashes:
+  ;;   a list of pre-hashed trees representing parameters to be curried into the puzzle.
+  ;;
+  ;; we return the hash of the curried expression
+  ;;   (a (q . mod-hash) (c (cp1 (c cp2 (c ... 1)...))))
+  ;;
+  ;; Note that from a user's perspective the hashes passed in here aren't simply
+  ;; the hashes of the desired parameters, but their treehash representation since
+  ;; that's the form we're assuming they take in the acutal curried program.
+
+  ;; inline functions that take varargs don't seem to work, so we can't inline `curry`
```

Comment:
You could inline it if you went `(defun curry (mod_hash_and_param_hashes)` then just use `(f` and `(r` to access the two args that are supposed to be there.  I'm pretty sure inlining will end up making a significant enough decrease in reveal size that it's worth it.

Reply:
I tried

```
  (defun-inline curry mod-hash-and-curry-parameter-hashes
     (tree-hash-of-apply (f mod-hash-and-curry-parameter-hashes)
                         (calculate-hash-of-curried-parameters (r mod-hash-and-curry-parameter-hashes)))
  )
```

and I can't get this to work either, likely because of compiler shortcomings.

Reply:
Have you tried in a file that `(include *standard-cl-21*)` is in?  It might be fine if it only works there since we should be using that going forward anyhow.  But otherwise yeah I don't see a way out.